### ISSUE TITLE
Make indices of IndexedJob pods configurable

### ISSUE THREAD

**Issue by isibeni** [kubernetes/kubernetes#109131]
### What would you like to be added?

As a user I would like to be able to configure with which index the pods of a IndexedJob are getting created. 

By introducing a new field to the job spec API (possible name: `indices` -> TBD) we can influence with which index the controller would create pods.

This field then has a mutually exclusive relation to the `completions` field.

Ideally, we could also have a new status field called `failedIndices` which contains the indices of the failed pods. 
,

### Why is this needed?

This is especially helpful when running large batch jobs where some of them are failing. 
The user would now have an easy interface to just recreate the failed job and only rerun the failed indices by specifying them via the newly introduced field.
If we also introduce the status field `failedIndices` the user could easily pick the failed indices from that status and recreate the job with those indices. Introducing this status field also allows automating this workflow

**Comment by alculquicondor** [comment]
cc @soltysh 

The field probably needs to be a string and it should have some limit for the number of characters.

As for the status, it might be harder to limit. In Indexed Jobs we limit `parallelism` to indirectly limit the number of characters in `completedIndexes`. We need to make a similar analysis to guarantee that the field doesn't grow indefinitely.

https://kubernetes.io/docs/reference/kubernetes-api/workload-resources/job-v1/#lifecycle

IMO, The feature would be a nice addition to Indexed Jobs. Let's see what others say.

**Comment by isibeni** [comment]
>The field probably needs to be a string and it should have some limit for the number of characters.

Agree. I would propose that it should have the same format as the `status.completedIndexes` field. e.g. `1,3-5,7` for 1, 3, 4, 5 and 7. 
If we then would also introduce a `status.failedIndexes` field a operator would then be able to just copy the string and insert it into the indexes field to retrigger a new job

**Comment by isibeni** [comment]
>In Indexed Jobs we limit parallelism to indirectly limit the number of characters in completedIndexes. We need to make a similar analysis to guarantee that the field doesn't grow indefinitely.

@alculquicondor Could you elaborate a bit what the problem with that is? Performance and memory consumption problems for scenarios where we can't simplify the string (e.g. all pods with even indexes are completing first)?

**Comment by alculquicondor** [comment]
Not just performance, but actually the etcd database has limits. You have identified the correct challenging scenario.

**Comment by isibeni** [comment]
> You have identified the correct challenging scenario.

And with introducing the new field the challenging scenario would be when every other index fails directly (e.g. `1,3,5,7,9,11,13,15`). We then would have the same value in `completedIndexes` and `failedIndexes`. Another one would be when every other index fails and every fourth index succeeds. `completedIndexes` could then contain the following value `1-3,5-7,9-11,13-15` while `failedIndexes` contains `1,3,5,7,9,11,13,15`.

Based on that we would need to check if we need to reduce the max value for `.spec.parallelism` even further

**Comment by alculquicondor** [comment]
If an index fails, but succeeds with a follow up pod, we shouldn't list it in `failedIndexes`.

For `completedIndexes`, `.spec.parallelism` limits how many non-continuous indexes can complete, because we create the indexes in order. Can the same rationale be applied for `failedIndexes`?

**Comment by isibeni** [comment]
sorry I just figured out that I had a small misconception about the behavior of backofflimit on indexed jobs.
I thought the backofflimit is applied for every index so that e.g. with a backofflimit of 2 each failing index will be restarted at least twice and if they reached the limit only those indexes will be considered as failed while the other indexes are allowed to run till completion. Still the overall job will be considered as failed. 
If indexed jobs would behave like that the field `failedIndexes` would make sense. But with the current implementation I don't see the sense of this field and tbh this proposed feature then also makes less sense. 

Or am I missing sth?

Maybe we can consider introducing a behaviour like a just described. This behaviour would also help with this issue: https://github.com/kubernetes/kubernetes/issues/107891

**Comment by alculquicondor** [comment]
The backofflimit is for the entire job.

I think the feature request still makes sense. The job overall is marked as Failed, but some indexes might have completed.

**Comment by alculquicondor** [comment]
> Maybe we can consider introducing a behaviour like a just described.

That would be a separate feature request. A we would need a new field, as we cannot change the behavior of existing fields, for backwards compatibility.

**Comment by isibeni** [comment]
> That would be a separate feature request. A we would need a new field, as we cannot change the behavior of existing fields, for backwards compatibility.

Agree on that.
Will open another issue these days


> I think the feature request still makes sense. The job overall is marked as Failed, but some indexes might have completed.

Yeah you are right. But then I would propose that we don't need the status field `failedIndexes` for now. This only becomes helpful with the other feature I just proposed

**Comment by soltysh** [comment]
This was brought to my attention during last wg-batch call. A few ideas worth considering. 

1. Do we actually need net-new field to express failed indices? Given current completions 100, for example, and a list of `completedIndexes` being 1-40,60-100 do we actually need that new filed if we should be able to calculate that using the existing fields?
2. I'd be curious to hear more about the use-cases for this particular request. @alculquicondor provided one during the call, but I'm not sure that's sufficient. I'd like to better understand the scenarios when and why a job (which by its definition is meant to run to completion) is failing and rather focus on improving the overall stability, reporting the failures or similar.

**Comment by ahg-g** [comment]
I created https://github.com/kubernetes/kubernetes/issues/109712 to track the idea of a job execution mode where every index is allowed to execute.

**Comment by alculquicondor** [comment]
/remove-lifecycle stale

We will have to tackle this in 1.27 or later.

**Comment by alculquicondor** [comment]
/lifecycle frozen
This is blocked on  #109712

**Comment by vsoch** [comment]
@alculquicondor what specifically is blocking? Why couldn't it be the case that you can configure the indices of the job, but not allow every index to execute?

**Comment by alculquicondor** [comment]
In a way, the features are independent. But doing #109712 first makes this feature more useful. So we started with #109712

The idea is that #109712 will include a field `failedIndexes: "1-5,7"`.
After you adjust some parameters that could have cause the failures, you can simply copy-paste "1-5,7" into a new Job that has a field `.spec.indexes: "1-5,7"`

**Comment by vsoch** [comment]
Gotcha - that does make sense. I'm very interested in this feature, and if they can be separate (and eventually integrated in the way you described) would it be OK if I started on it? I haven't contributed to Kubernetes proper and would like to try that out (so I'm more empowered to in the future). But if you have someone in mind and/or absolutely want to wait, I can respect that too! I was able to get a prototype for bursting with the Flux Operator, albeit I used a hack to get around this particular issue of not being able to control the indices. Let me know what you think, and if it's something I could help with, I would want to ask to be pointed to step 1. :)

**Comment by alculquicondor** [comment]
I don't think there were any takers previously. Note for a feature like this, you need to follow the KEP process: https://github.com/kubernetes/enhancements/tree/master/keps/sig-architecture/0000-kep-process

Ideally, before starting, it's helpful if you introduce yourself to SIG Apps (https://github.com/kubernetes/community/tree/master/sig-apps#meetings) announcing your intent and high level ideas, so that you clear possible questions and roadblocks in the future.
